src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -0,0 +1,303 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import java.util
+
+import com.fulcrumgenomics.FgBioDef._
+import htsjdk.samtools.{Cigar, CigarElement, SAMRecord, SAMUtils, CigarOperator => Op}
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+import scala.collection.mutable.ListBuffer
+
+/**
+  * Provides a suite of methods for clipping (soft- and hard-) bases from the beginnings
+  * and ends of reads in various ways.  Cigar strings, bases, qualities and alignment
+  * positions are all correctly adjusted post-clipping.
+  *
+  * Note that there are several "flavours" of method:
+  *   - Ones that work on the start and end of the read in whatever orientation the read
+  *     is in, vs. working on the 5' or 3' end of the read
+  *   - Ones that attempt to clip an additional N bases beyond any clipping already
+  *     provided (clip[Start|End|5PrimeEnd|3PrimeEnd]OfAlignment) and ones that
+  *     attempt to make it so that the read has N bases clipped, including any existing
+  *     clipping (clip[Start|End|5PrimeEnd|3PrimeEnd]OfRead)
+  */
+object SamRecordClipper {
+  /** An enumeration representing the various ways to clip bases within a read. */
+  object ClippingMode extends Enumeration { val Soft, Hard, SoftWithMask = Value }
+  type ClippingMode = ClippingMode.Value
+
+  private val NoCallBase = 'N'.toByte
+  private val NoCallQual = 2.toByte
+
+  /**
+    * Adds clipping of _at least_ numberOfBasesToClip to the start (left hand end)
+    * of an alignment. If clipping already exists at the start of the read it is
+    * preserved, and numberOfBasesToClip more clipping is added.
```

Comment:
I am surprised that we add more clipping rather than it being the desired # of bases to clip.  My thought was using these methods to ensure that all alignments had their first `N` bases clipped.